hydroflow_lang/src/graph/ops/join_fused.rs

```diff
@@ -0,0 +1,362 @@
+                _ => {
+                    diagnostics.push(Diagnostic::spanned(
+                        func.span(),
+                        Level::Error,
+                        format!("Unknown summarizing function: {func:?}"),
+                    ));
+                    return Err(());
+                }
+            }
+        }
+        _ => {
+            diagnostics.push(Diagnostic::spanned(
+                arg.span(),
+                Level::Error,
+                "argument must be a tuple",
+            ));
+            return Err(());
+        }
+    };
+
+    Ok(lhs)
+}
+
+pub fn make_joindata(
+    wc: &WriteContextArgs,
+    diagnostics: &mut Vec<Diagnostic>,
+    persistence: Persistence,
+    join_options: &JoinOptions<'_>,
+    side: &str,
+) -> Result<(Ident, Ident, TokenStream, TokenStream), ()> {
+    let joindata_ident = wc.make_ident(format!("joindata_{}", side));
+    let borrow_ident = wc.make_ident(format!("joindata_{}_borrow", side));
+
+    let context = wc.context;
+    let hydroflow = wc.hydroflow;
+    let root = wc.root;
+    let op_span = wc.op_span;
+
+    let join_type = match *join_options {
+        JoinOptions::FoldFrom(_, _) => {
+            quote_spanned!(op_span=> #root::compiled::pull::HalfJoinStateFoldFrom)
+        }
+        JoinOptions::Fold(_, _) => {
+            quote_spanned!(op_span=> #root::compiled::pull::HalfJoinStateFold)
+        }
+        JoinOptions::Reduce(_) => {
+            quote_spanned!(op_span=> #root::compiled::pull::HalfJoinStateReduce)
+        }
+    };
+
+    let (prologue, borrow) = match persistence {
+        Persistence::Tick => (
+            quote_spanned! {op_span=>
+                let #joindata_ident = #hydroflow.add_state(std::cell::RefCell::new(
+                    #root::util::monotonic_map::MonotonicMap::new_init(
+                        #join_type::default()
+                    )
+                ));
+            },
+            quote_spanned! {op_span=>
+                #borrow_ident.get_mut_clear(#context.current_tick())
+            },
+        ),
+        Persistence::Static => (
+            quote_spanned! {op_span=>
+                let #joindata_ident = #hydroflow.add_state(std::cell::RefCell::new(
+                    #join_type::default()
+                ));
+            },
+            quote_spanned! {op_span=>
+                #borrow_ident
+            },
+        ),
+        Persistence::Mutable => {
+            diagnostics.push(Diagnostic::spanned(
+                op_span,
+                Level::Error,
+                "An implementation of 'mutable does not exist",
+            ));
+            return Err(());
+        }
+    };
+    Ok((joindata_ident, borrow_ident, prologue, borrow))
+}
+
+pub fn parse_persistences(persistences: &[Persistence]) -> [Persistence; 2] {
+    match persistences {
+        [] => [Persistence::Tick, Persistence::Tick],
+        [a] => [*a, *a],
+        [a, b] => [*a, *b],
+        _ => unreachable!(),
```

Comment:
`pub(crate)`
And this line isn't really unreachable as a separate helper. `panic!`